cmd/expiration-mailer/main.go

```diff
@@ -0,0 +1,273 @@
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/mail"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+type emailContent struct {
+	ExpirationDate   time.Time
+	DaysToExpiration int
+	DNSNames         string
+}
+
+type mailer struct {
+	stats         statsd.Statter
+	log           *blog.AuditLogger
+	dbMap         *gorp.DbMap
+	mailer        mail.Mailer
+	emailTemplate *template.Template
+	nagTimes      []time.Duration
+	limit         int
+}
+
+func (m *mailer) sendNags(parsedCert *x509.Certificate, contacts []core.AcmeURL) error {
+	expiresIn := int(parsedCert.NotAfter.Sub(time.Now()).Hours()/24) + 1
+	emails := []string{}
+	for _, contact := range contacts {
+		if contact.Scheme == "mailto" {
+			emails = append(emails, contact.Opaque)
+		}
+	}
+	if len(emails) > 0 {
+		email := emailContent{
+			ExpirationDate:   parsedCert.NotAfter,
+			DaysToExpiration: expiresIn,
+			DNSNames:         strings.Join(parsedCert.DNSNames, ", "),
+		}
+		msgBuf := new(bytes.Buffer)
+		err := m.emailTemplate.Execute(msgBuf, email)
+		if err != nil {
+			m.stats.Inc("Mailer.Errors.SendingNag.TemplateFailure", 1, 1.0)
+			return err
+		}
+		err = m.mailer.SendMail(emails, msgBuf.String())
+		if err != nil {
+			m.stats.Inc("Mailer.Errors.SendingNag.SendFailure", 1, 1.0)
+			return err
+		}
+		m.stats.Inc("Mailer.Expiration.Sent", int64(len(emails)), 1.0)
+	}
+	return nil
+}
+
+func (m *mailer) updateCertStatus(serial string) error {
+	// Update CertificateStatus object
+	tx, err := m.dbMap.Begin()
+	if err != nil {
+		m.log.Err(fmt.Sprintf("Error opening transaction for certificate %s: %s", serial, err))
+		tx.Rollback()
+		return err
+	}
+
+	csObj, err := tx.Get(&core.CertificateStatus{}, serial)
+	if err != nil {
+		m.log.Err(fmt.Sprintf("Error fetching status for certificate %s: %s", serial, err))
+		tx.Rollback()
+		return err
+	}
+	certStatus := csObj.(*core.CertificateStatus)
+	certStatus.LastExpirationNagSent = time.Now()
+
+	_, err = tx.Update(certStatus)
+	if err != nil {
+		m.log.Err(fmt.Sprintf("Error updating status for certificate %s: %s", serial, err))
+		tx.Rollback()
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		m.log.Err(fmt.Sprintf("Error commiting transaction for certificate %s: %s", serial, err))
+		tx.Rollback()
+		return err
+	}
+
+	return nil
+}
+
+func (m *mailer) processCerts(certs []core.Certificate) {
+	m.log.Info(fmt.Sprintf("expiration-mailer: Found %d certificates, starting sending messages", len(certs)))
+	for _, cert := range certs {
+		regObj, err := m.dbMap.Get(&core.Registration{}, cert.RegistrationID)
+		if err != nil {
+			m.log.Err(fmt.Sprintf("Error fetching registration %d: %s", cert.RegistrationID, err))
+			m.stats.Inc("Mailer.Errors.GetRegistration", 1, 1.0)
+			continue
+		}
+		reg := regObj.(*core.Registration)
+		parsedCert, err := x509.ParseCertificate(cert.DER)
```

Comment:
Not to say we need to do this, but: If we extract the email addresses here, we can skip parsing the certificate if there are no contact addresses. Then still proceed on to update the status so we skip it next down the road.
Or we can filter in the SELECT, which may be a better choice.